### PR TITLE
Feature.primary group

### PR DESCRIPTION
**Pull request by indy-independence**
Every device is assigned exactly one primary group, and settings can be applied to this group. In the future maybe authorization can be controlled per primary group. A single device can still belong to multiple secondary groups, and secondary groups can be used to assign vxlans, sync settings, upgrade firmware etc.

settings repo: global/groups.yml syntax:
```
  - group:
      name: 'E1'
      regex: 'eosdist1$'
      group_priority: 100
  - group:
      name: 'E'
      regex: 'eosdist.*'
      group_priority: 99
```

device eosdist1 belongs to group E1, and eosdist2 belongs to group E.
Settings can then be set in settings repo groups/E/base_system.yml for example

### CODE REVIEW

File: src/cnaas_nms/db/settings.py [hunk 1]
```diff
@@ -396,6 +414,32 @@ def check_vlan_collisions(devices_dict: Dict[str, dict], mgmt_vlans: Set[int],
                 device_vlan_names[hostname] = {vxlan_data['vlan_name']}
 
 
+def check_group_priority_collisions(settings: Optional[dict] = None):
+    priorities: Dict[int, str] = {}
+    if not settings:
+        settings, _ = get_group_settings()
+    if settings is None:
```
Comment (pboers1988) [inline src/cnaas_nms/db/settings.py:421]: Twice the same check? Perhaps call get_group_settings before calling this function, and pass it along.

```python
Python 3.7.11 (default, Jul 27 2021, 07:03:16) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> settings = None
>>> not settings
True
>>> settings is None
True
>>> 
```

File: src/cnaas_nms/api/device.py [hunk 2]
```diff
@@ -183,12 +195,15 @@ def put(self, device_id):
...
 
             errors = dev.device_update(**json_data)
             if errors:
-                return empty_result(status='error', data=errors), 404
-            return empty_result(status='success', data={"updated_device": dev.as_dict()}), 200
+                return empty_result(status='error', data=errors), 400
+            session.commit()
+            update_device_primary_groups()
+            dev_dict = device_data_postprocess([dev])[0]
+            return empty_result(status='success', data={"updated_device": dev_dict}), 200
```
Comment (pboers1988) [inline src/cnaas_nms/api/device.py:206]: Perhaps a quirk I don't understand, but empty_result, does not seem that empty ;)

File: src/cnaas_nms/db/settings.py [hunk 3]
```diff
@@ -683,3 +751,91 @@ def get_group_regex(group_name: str) -> Optional[str]:
...
+    """Return dicts with {name: priority} for groups"""
+    groups_priorities = {}
+
+    if not settings:
+        settings, _ = get_group_settings()
+    if settings is None:
+        return groups_priorities
+    if 'groups' not in settings:
+        return groups_priorities
+    if settings['groups'] is None:
```
Comment (pboers1988) [inline src/cnaas_nms/db/settings.py:767]: The Above lines can be simplified:
```python
if not settings.get("groups", None):
    return group_priorities
```

File: src/cnaas_nms/db/settings.py [hunk 4]
```diff
@@ -683,3 +751,91 @@ def get_group_regex(group_name: str) -> Optional[str]:
...
+        return groups_priorities
+    for group in settings['groups']:
+        if 'name' not in group['group']:
```
Comment (pboers1988) [inline src/cnaas_nms/db/settings.py:770]: Seems like you are checking for all sorts of paramaters and skipping when certain keys don't exist? Perhaps it is better to not except the yaml file when parsing it, if there are syntax errors or keys missing? 

It seems to me that the iteration through the groups will be a lot easier.

Something like 

```python
for key, value in settings["group"].items():
    do the logic
```

File: src/cnaas_nms/db/settings.py [hunk 5]
```diff
@@ -683,3 +751,91 @@ def get_group_regex(group_name: str) -> Optional[str]:
...
+
+    Args:
+        no_cache: Update redis cache before returning data
+    """
+    logger = get_logger()
+    # update redis if redis is empty
+    with redis_session() as redis:
+        if not redis.exists("device_primary_group"):
+            no_cache = True
+    if no_cache:
```
Comment (pboers1988) [inline src/cnaas_nms/db/settings.py:832]: The lines above can be simplified:

```python
if not redis.exists("device_primary_group"):
    update_device_primary_groups()
```

**Review by pboers1988 — CHANGES_REQUESTED** [review]
If you take a look at the loops and if statements I'm pretty sure they can be simplified, by either calling certain functions sooner, or setting less variables, should make the code more readable and shorter.

File: src/cnaas_nms/db/settings.py (hunk 1 above)
Comment (indy-independence) [reply src/cnaas_nms/db/settings.py:421]: get_group_settings should only be called if you don't pass your own settings, so I don't want to call it before. The second check is if get_group_settings returns None, so I think they are both required actually, but I'll change so they both use the same syntax at least :)

File: src/cnaas_nms/api/device.py (hunk 2 above)
Comment (indy-independence) [reply src/cnaas_nms/api/device.py:206]: Haha yes indeed! A better name could maybe be make_response?

File: src/cnaas_nms/db/settings.py (hunk 3 above)
Comment (indy-independence) [reply src/cnaas_nms/db/settings.py:767]: Looks good!

File: src/cnaas_nms/db/settings.py (hunk 4 above)
Comment (indy-independence) [reply src/cnaas_nms/db/settings.py:770]: We should check the syntax of the yaml file but I think a lot of the stuff are optional, so there might not be a group defined at all etc

File: src/cnaas_nms/db/settings.py (hunk 5 above)
Comment (indy-independence) [reply src/cnaas_nms/db/settings.py:832]: Yeah ok that might be easier to read yes :)

**Pull request merged**